app/components/shared/inputs/TextInput.tsx

```diff
@@ -0,0 +1,66 @@
+import { Component, Prop } from 'vue-property-decorator';
+import cx from 'classnames';
+import styles from './TextInput.m.less';
+import { BaseInput } from './BaseInput';
+import { ITextMetadata } from './index';
+import { $t } from 'services/i18n';
+
+@Component({})
+export default class TextInput extends BaseInput<string, ITextMetadata> {
+  @Prop() readonly value: string;
+
+  @Prop({ default: () => ({}) })
+  readonly metadata: ITextMetadata;
+
+  textVisible = !this.metadata.masked;
+
+  toggleVisible() {
+    this.textVisible = !this.textVisible;
+  }
+
+  getValidations() {
+    return {
+      ...super.getValidations(),
+      date_format: this.options.dateFormat,
+      max: this.options.max,
+      min: this.options.min,
+      alpha_num: this.options.alphaNum,
+    };
+  }
+
+  toggleVisibleButton(h: Function) {
```

Comment:
Ouch so any function that returns XML needs `h` in context I guess?  I'm surprised there isn't a compiler to handle this for us.  Not a huge deal, just mildly annoying.

Reply:
yea that was actually something i was planning to look into-- might give it some time today